contracts/governance/StakingRewards/StakingRewards.sol

```diff
@@ -0,0 +1,152 @@
+ * Only wallets which have staked previously liquid SOV are eligible for these rewards.
+ * Tokenholders who stake their SOV receive staking rewards, a pro-rata share
+ * of the revenue that the platform generates from various transaction fees
+ * plus revenues from stakers who have a portion of their SOV slashed for
+ * early unstaking.
+ * */
+contract StakingRewards is StakingRewardsStorage, Initializable {
+	using SafeMath for uint256;
+
+	/// @notice Emitted when SOV is withdrawn
+	/// @param receiver The address which recieves the SOV
+	/// @param amount The amount withdrawn from the Smart Contract
+	event RewardWithdrawn(address indexed receiver, uint256 amount);
+
+	/**
+	 * @notice Replacement of constructor by initialize function for Upgradable Contracts
+	 * This function will be called only once by the owner.
+	 * @param _SOV SOV token address
+	 * @param _staking StakingProxy address should be passed
+	 * */
+	function initialize(address _SOV, IStaking _staking) external onlyOwner initializer {
+		require(_SOV != address(0), "Invalid SOV Address.");
+		require(Address.isContract(_SOV), "_SOV not a contract");
+		SOV = IERC20(_SOV);
+		staking = _staking;
+		startTime = block.timestamp;
+		setMaxDuration(26 * TWO_WEEKS);
+	}
+
+	/**
+	 * @notice Stops the current rewards program.
+	 * @dev All stakes existing on the contract at the point in time of
+	 * cancellation continue accruing rewards until the end of the staking
+	 * period being rewarded
+	 * */
+	function stop() external onlyOwner {
+		stopBlock = block.number;
+	}
+
+	/**
+	 * @notice Sets the max duration
+	 * @dev Rewards can be collected for a maximum duration at a time. This
+	 * is to avoid Block Gas Limit failures. Setting it zero would mean that it will loop
+	 * through the entire duration since the start of rewards program.
+	 * It should ideally be set to a value, for which the rewards can be easily processed.
+	 * @param _duration Max duration for which rewards can be collected at a go (in seconds)
+	 * */
+	function setMaxDuration(uint256 _duration) public onlyOwner {
+		maxDuration = _duration;
+	}
+
+	/**
+	 * @notice Collect rewards
+	 * @dev User calls this function to collect SOV staking rewards as per the SIP-0024 program.
+	 * The weighted stake is calculated using getPriorWeightedStake. Block number sent to the functon
+	 * must be a finalised block, hence we deduct 1 from the current block. User is only allowed to withdraw
+	 * after intervals of 14 days.
+	 * */
+	function collectReward() external {
+		uint256 count;
+		uint256 weightedStake;
+		uint256 lastFinalisedBlock = block.number - 1;
+		uint256 currentTS = block.timestamp;
+		address staker = msg.sender;
+		if (withdrawals[staker] == 0) {
+			require((currentTS.sub(startTime)) > TWO_WEEKS, "allowed after 14 days of start");
+			withdrawals[staker] = startTime;
+		} else {
+			require(currentTS > withdrawals[staker], "allowed after 14 days");
+		}
```

Comment:
It should not be 14 days since the last withdrawal or the start date, but instead a new staking interval must have started. There is a function on the staking contract:

```
timestampToLockDate(uint256 timestamp)
```

This function receives a time stamp and returns the last staking date prior to this date. E.g. staking date is the 7th of August, but now is the 9th of August, then this function returns the 7th of August.

This function is also used in the getPriorWeightedStake function. 

It would be cleaner to change the code here to work with these intervals.

```
lastStakingInterval = staking.timestampToLockDate(currentTS);
lastWithdrawalInterval = withdrawals[staker] > 0 ? withdrawals[staker] : startTime;
require(lastStakingInterval > lastWithdrawalInterval, "already claimed for the current interval");
```

startDate would need to be initilialized as timestampToLockDate(block.timestamp) in the constructor.
withdrawals[staker] would need to be set to lastStakingInterval before the reward is paid.